### PR TITLE
8270025: DynamicCallSiteDesc::withArgs doesn't throw NPE

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Please review this PR that is fixing a mismatch between the implementation for method `java.lang.constant.DynamicCallSiteDesc::withArgs` and its implementation. I made a mistake while working on a recent CSR [JDK-8224985](https://bugs.openjdk.java.net/browse/JDK-8224985) and fixed the API but mistakenly thought that the implementation was in sync with the spec. This is why this change is also including a unit test of the API for `java.lang.constant.DynamicCallSiteDesc` modulo method `resolveCallSiteDesc` which is covered in test `IndyDescTest` in the same test suite. Also this change needs a CSR as while fixing the implementation of method `::withArgs` I realized that the API of the varargs overloaded version of method `::of` needed some rewording too as it is invoking the same private constructor `::withArgs` is invoking. So it didn't make sense for the API of one method to be more restrictive than the other. Please review also the accompanying CSR.

Thanks,
Vicente

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270025](https://bugs.openjdk.java.net/browse/JDK-8270025): DynamicCallSiteDesc::withArgs doesn't throw NPE

### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to 602a85c4612cffd4191d4bd55c092ef8136aabd8
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/242/head:pull/242` \
`$ git checkout pull/242`

Update a local copy of the PR: \
`$ git checkout pull/242` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/242/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 242`

View PR using the GUI difftool: \
`$ git pr show -t 242`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/242.diff">https://git.openjdk.java.net/jdk17/pull/242.diff</a>

</details>

### CODE REVIEW

**Review by JornVernee — APPROVED** [review]
LGTM

File: test/jdk/java/lang/constant/DynamicCallSiteDescTest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+     */
+
+    public void testOf() throws ReflectiveOperationException {
+        try {
+            DynamicCallSiteDesc.of(
+                    ConstantDescs.ofCallsiteBootstrap(
+                            ClassDesc.of("BootstrapAndTarget"),
+                            "bootstrap",
+                            ClassDesc.of("java.lang.invoke.CallSite")
+                    ),
```
Comment (mlchung) [inline test/jdk/java/lang/constant/DynamicCallSiteDescTest.java:60]: A minor suggestion: you can have the return `DirectMethodHandleDesc` in a local variable to be used in all `DynamicCallSiteDesc::of` calls that would avoid copy-n-paste of same `ConstantDescs::ofCallsiteBootstrap` call.

File: test/jdk/java/lang/constant/DynamicCallSiteDescTest.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                    "",
+                    MethodTypeDesc.ofDescriptor("()I")
+            );
+            throw new AssertionError("IllegalArgumentException expected");
```
Comment (mlchung) [inline test/jdk/java/lang/constant/DynamicCallSiteDescTest.java:64]:
```suggestion
            fail("IllegalArgumentException expected");
```

Same for other `throw new AssertionError(...)`

**Review by mlchung — APPROVED** [review]
Looks good.  Minor suggestions in the test.

File: test/jdk/java/lang/constant/DynamicCallSiteDescTest.java (hunk 1 above)
Comment (vicente-romero-oracle) [reply test/jdk/java/lang/constant/DynamicCallSiteDescTest.java:60]: will do

**Review by mlchung — APPROVED** [review]

**Comment by vicente-romero-oracle** [comment]
thanks for the reviews!

**Pull request closed without merging**